crypto/crypto/src/unit_tests/noise_test.rs

```diff
@@ -9,8 +9,27 @@ use crate::{
     x25519, Uniform as _,
 };
 
+use crate::ed25519::{Ed25519PrivateKey, Ed25519PublicKey};
 use rand::SeedableRng;
 use serde::*;
+use std::convert::TryFrom;
+
+#[test]
+fn convert_from_ed25519_publickey() {
```

Comment:
This test will fail on some inputs, specifically on inputs that are not [X25519-canonical](https://tools.ietf.org/html/rfc7748#section-5). I reimplemented it as a proptest on top of my #3904 here:

https://github.com/huitseeker/libra/compare/crypto-shenanigans...huitseeker:this_test_doesnt_work?expand=1

The test output, including a counterexample:
https://gist.github.com/1c93555d176f731af485f740a92c8615

Reply:
Yes we could add this test here as well. I do catch it on try_from, but didn't test it here.

Reply:
@huitseeker note that in your referenced test, you use the Ed private key bytes directly (w/o expansion)
`let x25519_privatekey = x25519::PrivateKey::try_from(&keypair.private_key.to_bytes()[..]).unwrap();` which will always fail.

, while in my approach I produce the expanded key first (after sha512), then Ed25519 key clamping, then get the first 32bytes. Based on how clamping of the most significant bits works on Ed25519 and x25519 (and because Ed25519 is more restrictive as it sets to zero the 2nd most significant bit), this conversion is always possible.

However, I get it that this does not conform 100% w/ the x25519 standard (due to this 2nd bit twiddling), but this is the price to pay if we want to support what @davidiw expects here.

Reply:
Oh yes indeed! :facepalm: Yeah, I messed up the test. But you can still work around the bit-clamping of X25519 by rejecting inputs where it has to apply.

I also have the conversion in my PR (`ed25519::Ed25519PrivateKey::from_ed25519_private_bytes`)
and I outline the difference in type of conversions in the tests `convert_from_ed25519_publickey` and `ed25519_and_x25519_privkeys`. 

In the former, you take in ed25519 bits, and you *reinterpret* what is the "moral" key in there, i.e. the scalar that links the public key to the private key. I reject those scalars that cannot be represented as x25519 scalars unmodified, because in our block chain a signature verification failure can have dire consequences (e.g. if your public noise signature is not what's expected, you may be considered as an atacker). That's one approach I call "reinterpretation" for the sake of discussion.

In `ed25519_and_x25519_privkeys` I consider you have marshalled an x25519 keypair inside an ed25519 storage device, knowing what you're doing. In that case the public and private key ed25519 key do not "match" each other, but the x25519 keypair you can extract from that is valid in every case. That's what I call marshalling.

As discussed with @valerini over review, both approaches are valid, and the inputs to them are genuinely untyped: without user input, there really is two ways to see an ed25519 key pair as x25519. Hence why I now think more strongly that a serialization/deserialization approach works better than a conversion.